File: kirun-js/src/engine/function/system/Make.ts
import { AbstractFunction } from '../AbstractFunction';
import { FunctionSignature } from '../../model/FunctionSignature';
import { FunctionExecutionParameters } from '../../runtime/FunctionExecutionParameters';
import { FunctionOutput } from '../../model/FunctionOutput';
import { Namespaces } from '../../namespaces/Namespaces';
import { Event } from '../../model/Event';
import { Parameter } from '../../model/Parameter';
import { Schema } from '../../json/schema/Schema';
import { EventResult } from '../../model/EventResult';
import { MapUtil } from '../../util/MapUtil';
import { ExpressionEvaluator } from '../../runtime/expression/ExpressionEvaluator';
import { TokenValueExtractor } from '../../runtime/expression/tokenextractor/TokenValueExtractor';

const RESULT_SHAPE = 'resultShape';
const VALUE = 'value';

export class Make extends AbstractFunction {
    private readonly signature: FunctionSignature;

    public constructor() {
        super();
        this.signature = new FunctionSignature('Make')
            .setNamespace(Namespaces.SYSTEM)
            .setParameters(
                new Map([Parameter.ofEntry(RESULT_SHAPE, Schema.ofAny(RESULT_SHAPE))]),
            )
            .setEvents(
                new Map([
                    Event.outputEventMapEntry(MapUtil.of(VALUE, Schema.ofAny(VALUE))),
                ]),
            );
    }

    public getSignature(): FunctionSignature {
        return this.signature;
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        const args = context.getArguments() ?? new Map<string, any>();
        const resultShape = args.get(RESULT_SHAPE);
        const valuesMap: Map<string, TokenValueExtractor> = context.getValuesMap();

        const result = this.processValue(resultShape, valuesMap);

        return new FunctionOutput([EventResult.outputOf(MapUtil.of(VALUE, result))]);
    }

    private processValue(value: any, valuesMap: Map<string, TokenValueExtractor>): any {
        if (value === null || value === undefined) {
            return value;
        }

        if (typeof value === 'string') {
            return this.evaluateExpression(value, valuesMap);
        }

        if (Array.isArray(value)) {
            return value.map((item) => this.processValue(item, valuesMap));
        }

        if (typeof value === 'object') {
            const result: Record<string, any> = {};
            for (const [key, val] of Object.entries(value)) {
                result[key] = this.processValue(val, valuesMap);
            }
            return result;
        }

        return value;
    }

    private evaluateExpression(expression: string, valuesMap: Map<string, TokenValueExtractor>): any {
        if (!expression || !expression.startsWith('{{') || !expression.endsWith('}}')) {
            return expression;
        }

        const innerExpression = expression.substring(2, expression.length - 2);
        const evaluator = new ExpressionEvaluator(innerExpression);
        const result = evaluator.evaluate(valuesMap);

        return result !== undefined ? result : null;
    }
}

File: kirun-py/src/kirun_py/function/system/make.py
from __future__ import annotations

from typing import Any, Dict, Optional, TYPE_CHECKING

from kirun_py.function.abstract_function import AbstractFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event import Event
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.model.function_signature import FunctionSignature
from kirun_py.model.parameter import Parameter
from kirun_py.namespaces.namespaces import Namespaces
from kirun_py.runtime.expression.expression_evaluator import ExpressionEvaluator
from kirun_py.runtime.expression.tokenextractor.token_value_extractor import TokenValueExtractor

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

RESULT_SHAPE = 'resultShape'
VALUE = 'value'


class Make(AbstractFunction):

    def __init__(self) -> None:
        super().__init__()
        self._signature = (
            FunctionSignature('Make')
            .set_namespace(Namespaces.SYSTEM)
            .set_parameters(dict([
                Parameter.of_entry(RESULT_SHAPE, Schema.of_any(RESULT_SHAPE)),
            ]))
            .set_events(dict([
                Event.output_event_map_entry({VALUE: Schema.of_any(VALUE)}),
            ]))
        )

    def get_signature(self) -> FunctionSignature:
        return self._signature

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        args = context.get_arguments() if context.get_arguments() is not None else {}
        result_shape = args.get(RESULT_SHAPE)
        values_map: Dict[str, TokenValueExtractor] = context.get_values_map()

        result = self._process_value(result_shape, values_map)

        return FunctionOutput([EventResult.output_of({VALUE: result})])

    def _process_value(
        self, value: Any, values_map: Dict[str, TokenValueExtractor]
    ) -> Any:
        if value is None:
            return value

        if isinstance(value, str):
            return self._evaluate_expression(value, values_map)

        if isinstance(value, list):
            return [self._process_value(item, values_map) for item in value]

        if isinstance(value, dict):
            return {
                key: self._process_value(val, values_map)
                for key, val in value.items()
            }

        return value

    def _evaluate_expression(
        self, expression: str, values_map: Dict[str, TokenValueExtractor]
    ) -> Any:
        if not expression or not expression.startswith('{{') or not expression.endswith('}}'):
            return expression

        inner_expression = expression[2:-2]
        evaluator = ExpressionEvaluator(inner_expression)
        result = evaluator.evaluate(values_map)

        return result if result is not None else None
